src/v/v8_engine/executor.cc

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "v8_engine/executor.h"
+
+#include "seastarx.h"
+#include "utils/mutex.h"
+
+#include <seastar/core/alien.hh>
+#include <seastar/core/do_with.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/semaphore.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/core/smp.hh>
+#include <seastar/util/later.hh>
+
+#include <chrono>
+#include <memory>
+
+namespace v8_engine {
+
+namespace internal {
+
+// spsc_queue
+
+spsc_queue::spsc_queue()
+  : _items(ss::smp::count)
+  , _ss_mutex_shard(ss::this_shard_id()) {}
+
+void spsc_queue::close() {
+    is_stopped = true;
+    _push_mutex.broken(ss::gate_closed_exception());
+    _has_element_cv.notify_all();
+}
+
+ss::future<> spsc_queue::push(work_item* item) {
+    auto lock = co_await _push_mutex.get_units();
+
+    while (!_items.push(item)) {
+        co_await _is_not_full.wait();
+    }
+
+    _has_element_cv.notify_one();
+    co_return;
+}
+
+work_item* spsc_queue::pop() {
+    work_item* item = nullptr;
+    std::unique_lock lock{_std_mutex};
+
+    // We need to use wait_for, because std::thread can lose exception from
```

Comment:
should the comment say:  `loose notification` ?

Reply:
done